Utilities/XrdAdaptor/src/XrdRequestManager.cc

```diff
@@ -370,10 +370,16 @@ RequestManager::checkSourcesImpl(timespec &now, IOSize requestSize)
         }
     }
   }
+  // NOTE: OpenHandler will take an internal mutex, then invoke RequestManager methods that take
+  // m_source_mutex.  This ordering is also done within OpenHandler::HandleResponseWithHosts; hence,
+  // we must release m_source_mutex before calling open() or deadlock due to mutex ordering.
   if (findNewSource)
   {
-    m_open_handler->open();
     m_lastSourceCheck = now;
+    sentry.unlock();
+    // Automatically reacquire the lock after this code block finishes.
+    std::shared_ptr<const void> reacquire_lock_sentry(nullptr, [&sentry](void*) {sentry.lock();});
+    m_open_handler->open();
```

Comment:
Turns out this is not safe. There is a comment in OpenHandler::open() which explicitly states that this function isn't thread safe and requires a lock on it to guarantee you can't call it from multiple threads. The problem is an unguarded access to a member data.